### ISSUE TITLE
The 2027 problem

### ISSUE THREAD

**Issue by atsampson** [PDP-10/its#853]
From MC:SYSDOC;ITS RECENT:

> Date: Mon, 26 Oct 87 11:58:03 EST
> From: Alan Bawden
> [...]
> The other, more common, date format is the one used by the filesystem, and by programs that used the DATIME library; this is the format returned by the RQDATE system call.  This format packs the year, less 1900, in a 7-bit field, so this will last through 2027.  Since the bit to the left of the year field is unused, we can easily expand this to last through 2155.

It's looking like ITS will be alive and well come 2027, so this'll need fixing at some point.

Things that will need fixing (very incomplete!):

- [ ] SYSTEM
- [ ] SYSDOC; .CALLS
- [ ] TJ6
- [ ] TECO's FS FDCONV

**Comment by larsbrinkhoff** [comment]
I looked at using eight or nine bits for the year field.  For eight, it's generally agreed that it's probably no problem.  For nine, I didn't see anything obviously damaging, but who knows.

Anyone volunteering to run a system with the clock set 300 yeas ahead?

**Comment by atsampson** [comment]
The system's idea of the date is stored in QDATE, which is updated by CLCQDT (in TIME) and read along with TIMOFF by various syscalls in DISK (e.g. ARQDAT for .RQDATE). It looks like it would be straightforward to make CLCQDT write the year into 4.9-4.1 rather than 4.7-4.1.

The filesystem doesn't appear to use bits 4.8 or 4.9 in UNDATE or UNREF for anything.

However, QDATE is set to -1 on boot to indicate that the time isn't known yet, and there are a few tests that do things like `SKIPL A,QDATE`. These would need adjusting to test for -1 specifically if we wanted to use 4.9.

**Comment by larsbrinkhoff** [comment]
Thanks for checking.  There's also a chance some user program may use a negative value to signal a special condition.

**Comment by atsampson** [comment]
SYSDOC; .CALLS > will need updating for this; at the same time, we can fix the various places where it says that the year field is "mod 100" when it's actually "minus 1900".

**Comment by larsbrinkhoff** [comment]
I set my TT ITS clock to 2028 and made a file with CREATE.  It got timestamped 1928.

**Comment by larsbrinkhoff** [comment]
Must have clicked the wrong button.

**Comment by larsbrinkhoff** [comment]
The TV-11 code claims it's limited to a 7-bit year.

**Comment by larsbrinkhoff** [comment]
SYSENG; DATIME 74 has this:

```
TM%YR==  177000,,0 ; 4.7-4.1    0-127   years   0-127 relative to 1900 (1900-2027)

**Comment by larsbrinkhoff** [comment]
Year 2028 is not far off.  Guess it's time to get started on this!

**Comment by larsbrinkhoff** [comment]
Another new year, closer to 2028.

**Comment by larsbrinkhoff** [comment]
I have actually started something on this branch: https://github.com/PDP-10/its/tree/lars/y2038

View the diff here: https://github.com/PDP-10/its/compare/master...lars/y2038

**Comment by larsbrinkhoff** [comment]
Things to audit:

- [x] Use of QDATE inside ITS.
- [x] Use of file timestamps inside ITS.
- [ ] Do CDATE/RDATE (UNDATE/UNREF) really use "all ones" as "date unknown"?
- [x] Use of RQDATE for handling current time in applications.
- [x] Use of sign bit from RQDATE to check for valid date.
- [x] Use of byte pointers `330700,,` or masks `177000,,`.
- [x] Use of RFDATE, SFDATE, SRDATE, and FILBLK for handling file timestamps in applications.
- [ ] Use of .RDATE, .RDATIM, .RLPDTM, and .RYEAR in applications.
- [ ] Leap years 2000 and 2400.

**Comment by larsbrinkhoff** [comment]
Curiously I have encountered a few places where the byte pointer LH is 330800 or 330900.  Is that valid?!

EDIT: Yes, they are valid.  Both MIDAS and DDT accept that syntax and do the right thing.

Searching around, I also see many cases of 440800.  I guess it's a stylistic thing that can be retained rather than "fixed".

**Comment by eswenson1** [comment]
I assume midas will just turn the 330800 into a 331000 and a 330900 into a 331100?  And therefore it will move 8 (or 9) bits into our out of bit positions 0-8 (9 bits) of the addressed word?

**Comment by larsbrinkhoff** [comment]
Yes, 330800 and 330900 turn into 331000 and 331100.  As the left half of a byte pointer, they access the left end (most significant bits) of the addressed word, i.e.  1-8 or 0-8 in big endian bit notation.  That's where the "disk format" date stores the year.  (Originally it was 2-8, 330700.)

I guess ITS hackers felt it natural to use 8 or 9 to refer to those size bytes, even thought it's part of an octal number.  That's fine, as long as you don't try to use 10 to get a 10-bit byte!

**Comment by larsbrinkhoff** [comment]
This aptly illustrates the urgency of the update.

<img width="676" height="445" alt="Image" src="https://github.com/user-attachments/assets/f0213f7b-f203-4708-ba6a-8d174d213f6f" />

**Comment by eswenson1** [comment]
I assume you meant: 1-8 or 0-7, right?

**Comment by larsbrinkhoff** [comment]
No.  1-8 is byte pointer 331000,, and 0-8 is byte pointer 331100,,

**Comment by eswenson1** [comment]
Gotcha. I thought you were providing two equivalent bit ranges — one 0-based and the other 1-based!
